disk_objectstore/backup_utils.py

```diff
@@ -0,0 +1,382 @@
+        return self.run_cmd(cmd)[0]
+
+    def validate(self):
+        """Validate the backup config inputs"""
+        if self.keep < 0:
+            raise BackupError(
+                "Input validation failed: keep variable can't be negative!"
+            )
+
+        if self.remote:
+            self.check_if_remote_accessible()
+
+        if self.exes:
+            for _, path in self.exes.items():
+                if not is_exe_found(path):
+                    raise BackupError(
+                        f"Input validation failed: {path} not accessible."
+                    )
+
+        path_exists = self.check_path_exists(self.path)
+
+        if not path_exists:
+            success = self.run_cmd(["mkdir", str(self.path)])[0]
+            if not success:
+                raise BackupError(
+                    f"Input validation failed: Couldn't access/create '{str(self.path)}'!"
+                )
+
+    def run_cmd(
+        self,
+        args: list,
+    ):
+        """
+        Run a command locally or remotely.
+        """
+        all_args = args[:]
+        if self.remote:
+            all_args = ["ssh", self.remote] + all_args
+
+        res = subprocess.run(all_args, capture_output=True, text=True, check=False)
+
+        self.logger.debug(
+            f"Command: {all_args}\n"
+            f"  Exit Code: {res.returncode}\n"
+            f"  stdout/stderr: {res.stdout}\n{res.stderr}"
+        )
+
+        success = res.returncode == 0
+
+        return success, res.stdout
+
+    def call_rsync(  # pylint: disable=too-many-arguments
+        self,
+        src: Path,
+        dest: Path,
+        link_dest: Optional[Path] = None,
+        src_trailing_slash: bool = False,
+        dest_trailing_slash: bool = False,
+        extra_args: Optional[list] = None,
+    ):
+        """Call rsync with specified arguments and handle possible errors & stdout/stderr
+
+        :param link_dest:
+            Path to the hardlinked files location (previous backup).
+
+        :param src_trailing_slash:
+            Add a trailing slash to the source path. This makes rsync copy the contents
+            of the folder instead of the folder itself.
+
+        :param dest_trailing_slash:
+            Add a trailing slash to the destination path. This makes rsync interpret the
+            destination as a folder and create it if it doesn't exists.
+
+        """
+
+        assert "rsync" in self.exes
+
+        all_args = [self.exes["rsync"], "-azh", "-vv", "--no-whole-file"]
+        if extra_args:
+            all_args += extra_args
+        if link_dest:
+            if not self.remote:
+                # for local paths, use resolve() to get absolute path
+                link_dest_str = str(link_dest.resolve())
+            else:
+                # for remote paths, we require absolute paths anyways
+                link_dest_str = str(link_dest)
+            all_args += [f"--link-dest={link_dest_str}"]
```

Comment:
No need to explicitly call `str()` this is done automatically by the f-string
```suggestion
                link_dest = link_dest.resolve()
            all_args += [f"--link-dest={link_dest}"]
```